cmd/pj-rehearse/main.go

```diff
@@ -1,23 +1,278 @@
+				c.logger.WithFields(logFields).Info("Rehearsal job uses ci-operator config ConfigMap")
+			}
+		}
+	}
+}
+
+// Create a rehearsal ConfigMap with ci-operator config files needed by the
+// rehearsal jobs.
+func (c *rehearsalCIOperatorConfigs) Create() error {
+	cm := &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{Name: c.configMapName},
+		Data:       map[string]string{},
+	}
+	c.logger.Debug("Preparing rehearsal ConfigMap for ci-operator configs")
+
+	for key, path := range c.neededConfigs {
+		fullPath := filepath.Join(c.configDir, path)
+		content, err := ioutil.ReadFile(fullPath)
+		c.logger.WithField("ciop-config", key).Info("Loading ci-operator config to rehearsal ConfigMap")
+		if err != nil {
+			return fmt.Errorf("failed to read ci-operator config file from %s: %v", fullPath, err)
+		}
+
+		cm.Data[key] = string(content)
+	}
+
+	if c.dry {
+		cmAsYAML, err := yaml.Marshal(cm)
+		if err != nil {
+			return fmt.Errorf("Failed to marshal ConfigMap to YAML: %v", err)
+		}
+		fmt.Printf("%s\n", cmAsYAML)
+		return nil
+	}
+	c.logger.Info("Creating rehearsal ConfigMap for ci-operator configs")
+	_, err := c.cmclient.Create(cm)
+	return err
+}
+
+func makeRehearsalPresubmit(source *prowconfig.Presubmit, repo string, prNumber int) (*prowconfig.Presubmit, error) {
+	var rehearsal prowconfig.Presubmit
+	deepcopy.Copy(&rehearsal, source)
+
+	rehearsal.Name = fmt.Sprintf("rehearse-%d-%s", prNumber, source.Name)
+	rehearsal.Context = fmt.Sprintf("ci/rehearse/%s/%s", repo, strings.TrimPrefix(source.Context, "ci/prow/"))
+
+	if len(source.Spec.Containers) != 1 {
+		return nil, fmt.Errorf("Cannot rehearse jobs with more than 1 container in Spec")
+	}
+	container := source.Spec.Containers[0]
+
+	if len(container.Command) != 1 || container.Command[0] != "ci-operator" {
+		return nil, fmt.Errorf("Cannot rehearse jobs that have Command different from simple 'ci-operator'")
+	}
+
+	for _, arg := range container.Args {
+		if strings.HasPrefix(arg, "--git-ref") || strings.HasPrefix(arg, "-git-ref") {
+			return nil, fmt.Errorf("Cannot rehearse jobs that call ci-operator with '--git-ref' arg")
+		}
+	}
+
+	if len(source.Branches) != 1 {
+		return nil, fmt.Errorf("Cannot rehearse jobs that run over multiple branches")
+	}
+	branch := strings.TrimPrefix(strings.TrimSuffix(source.Branches[0], "$"), "^")
+
+	gitrefArg := fmt.Sprintf("--git-ref=%s@%s", repo, branch)
+	rehearsal.Spec.Containers[0].Args = append(source.Spec.Containers[0].Args, gitrefArg)
+
+	return &rehearsal, nil
+}
+
+func submitRehearsal(job *prowconfig.Presubmit, jobSpec *pjapi.ProwJobSpec, logger logrus.FieldLogger, pjclient pj.ProwJobInterface, dry bool) (*pjapi.ProwJob, error) {
+	labels := make(map[string]string)
+	for k, v := range job.Labels {
+		labels[k] = v
+	}
+
+	pj := pjutil.NewProwJob(pjutil.PresubmitSpec(*job, *(jobSpec.Refs)), labels)
+	logger.WithFields(pjutil.ProwJobFields(&pj)).Info("Submitting a new prowjob.")
+
+	if dry {
```

Comment:
Can we do this in a wrapper of the client?

Reply:
I guess we can, for both clients we currently use for modifying calls.